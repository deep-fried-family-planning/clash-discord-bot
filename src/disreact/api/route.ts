import type {Ix} from '#src/internal/disreact/virtual/entities/dapi.ts';



export type DisReactRoute = {
  root  : string;
  node  : string;
  search: URLSearchParams;
};


export const decodeFromURL = (url: URL): DisReactRoute => {
  return {
    root  : '',
    node  : '',
    search: new URLSearchParams(url.searchParams),
  };
};


export const decodeFromPathname = (pathname: string): DisReactRoute => {
  const [root, node] = pathname.split('/');
  const search       = new URLSearchParams();

  return {
    root,
    node,
    search,
  };
};


export const getInteractionRoutingInfo = (rest: Ix): DisReactRoute => {

};
