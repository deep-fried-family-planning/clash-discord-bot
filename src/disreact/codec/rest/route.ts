import {SnowFlake} from '#src/disreact/codec/constants/common.ts';
import {Dokens, Event} from '#src/disreact/codec/rest/index.ts';
import type {Schema} from 'effect/Schema';
import {Any, mutable, Struct, Union} from 'effect/Schema';
import {DialogParams, EmbedParams, Params} from 'src/disreact/codec/rest/index.ts';



export const T = mutable(Struct({
  request: Any,
  id     : SnowFlake,
  params : Union(DialogParams.T, EmbedParams.T),
  dokens : Dokens.T,
  event  : Event.T,
}));

export type T = Schema.Type<typeof T>;

export const decodeRouteFromRequest = (request: any): T => {
  const dokens = Dokens.make(request);
  const params = Params.decodeFromRequest(request);
  const event  = DialogParams.is(params)
    ? Event.decodeRequestEvent(request, params.custom_id)
    : Event.decodeRequestEvent(request);

  return {
    request,
    id    : request.id,
    params,
    dokens: Dokens.add(dokens, params),
    event,
  };
};
