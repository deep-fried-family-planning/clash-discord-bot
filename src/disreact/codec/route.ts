import {S} from '#src/disreact/codec/re-exports.ts'

export * as Route from './route.ts'
export type Route = never

export const Message = S.Struct({
  id: S.String,
})
