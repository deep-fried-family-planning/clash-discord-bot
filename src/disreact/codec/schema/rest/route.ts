import {DateTimeUtcFromNumber, mutable, NumberFromString, String, Struct, TemplateLiteralParser} from 'effect/Schema';
import * as Doken from './doken.ts';



export const MessageParams = Struct({
  root : String,
  doken: Doken.Type,
  hash : String,
});



export const MessageParser = TemplateLiteralParser(
  '/dsx/',
  String,
  '/',
  String,
  '/',
  NumberFromString,
  '/',
  String,
  '/',
  String,
  '/',
  String,
);
