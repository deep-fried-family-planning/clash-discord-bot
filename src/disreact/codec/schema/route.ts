import {Rest} from '#src/disreact/codec/abstract/index.ts';
import {Struct, String, mutable, optional, NumberFromString, DateTimeUtcFromNumber, TemplateLiteralParser} from 'effect/Schema';



export const MainParameters = Struct({
  root : String,
  doken: mutable(Struct({
    app  : String,
    id   : String,
    token: String,
    ttl  : DateTimeUtcFromNumber,
    type : NumberFromString,
    flags: NumberFromString,
  })),
  hash: String,
});



export const MainSerialized = TemplateLiteralParser(
  '/dsx/',
  String,
  '/',
  String,
  '/',
  NumberFromString,
  '/',
  String,
  '/',
  String,
  '/',
  String,
);
