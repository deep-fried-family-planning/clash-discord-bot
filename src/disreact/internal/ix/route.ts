export type T = {
  root: string;
};
