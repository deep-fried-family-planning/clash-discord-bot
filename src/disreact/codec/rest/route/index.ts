export * as DialogParams from '#src/disreact/codec/rest/route/params/dialog-params.ts';
export * as EmbedParams from '#src/disreact/codec/rest/route/params/embed-params.ts';
export * as Event from 'src/disreact/codec/rest/route/event.ts';
import {SnowFlake} from '#src/disreact/codec/constants/common.ts';
import {CLICK} from '#src/disreact/codec/rest/rest.ts';
import * as Dokens from '#src/disreact/codec/rest/route/dokens.ts';
import * as Event from '#src/disreact/codec/rest/route/event.ts';
import * as DialogParams from '#src/disreact/codec/rest/route/params/dialog-params.ts';
import * as EmbedParams from '#src/disreact/codec/rest/route/params/embed-params.ts';
import {Any, mutable, type Schema, Struct, Union} from 'effect/Schema';



export const Route = mutable(Struct({
  request: Any,
  id     : SnowFlake,
  params : Union(DialogParams.DialogParams, EmbedParams.T),
  dokens : Dokens.Dokens,
  event  : Event.Event,
}));

export type Route = Schema.Type<typeof Route>;

export const decodeRequestRoute = (request: any): Route => {
  const dokens = Dokens.makeDokens(request);

  if (request.type === CLICK) {
    const params = EmbedParams.decodeMessageRouteFromRequest(request)!;
    const event  = Event.decodeRequestEvent(request);
    dokens.defer = params.doken;

    return {
      request,
      id: request.id,
      params,
      dokens,
      event,
    };
  }

  const params = DialogParams.decodeDialogParams(request.data.custom_id, request);
  const event  = Event.decodeRequestEvent(request, params.custom_id);
  dokens.defer = params.message?.doken;

  return {
    request,
    id: request.id,
    params,
    dokens,
    event,
  };
};
