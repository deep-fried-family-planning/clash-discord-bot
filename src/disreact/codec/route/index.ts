import {SnowFlake} from '#src/disreact/codec/constants/common.ts';
import {CLICK} from '#src/disreact/codec/rest/rest.ts';
import * as Dokens from '#src/disreact/codec/route/dokens.ts';
import * as Event from '#src/disreact/codec/route/event.ts';
import {Any, mutable, type Schema, Struct, Union} from 'effect/Schema';
import * as DialogParams from './dialog-params.ts';
import * as MessageRoute from './message-route.ts';

export * from './dialog-params.ts';
export * from './message-route.ts';
export * as Event from './event.ts';

export const Route = mutable(Struct({
  request: Any,
  id     : SnowFlake,
  params : Union(DialogParams.DialogParams, MessageRoute.MessageRoute),
  dokens : Dokens.Dokens,
  event  : Event.Event,
}));

export type Route = Schema.Type<typeof Route>;

export const decodeRequestRoute = (request: any): Route => {
  const dokens = Dokens.makeDokens(request);

  if (request.type === CLICK) {
    const params = MessageRoute.decodeMessageRouteFromRequest(request)!;
    const event  = Event.decodeRequestEvent(request);
    dokens.defer = params.doken;

    return {
      request,
      id: request.id,
      params,
      dokens,
      event,
    };
  }

  const params = DialogParams.decodeDialogParams(request.data.custom_id, request);
  const event  = Event.decodeRequestEvent(request, params.custom_id);
  dokens.defer = params.message?.doken;

  return {
    request,
    id: request.id,
    params,
    dokens,
    event,
  };
};
