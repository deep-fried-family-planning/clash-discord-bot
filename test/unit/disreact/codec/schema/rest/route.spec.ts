import {E, L, RDT} from '#src/internal/pure/effect.ts';
import {layer} from '@effect/vitest';
import {DateTime, Logger, LogLevel, pipe, TestContext} from 'effect';
import * as Route from '#src/disreact/codec/rest/route.ts';



const describe = pipe(
  L.empty,
  L.provideMerge(TestContext.TestContext),
  L.provideMerge(Logger.replace(Logger.defaultLogger, Logger.prettyLoggerDefault)),
  L.provideMerge(Logger.minimumLogLevel(LogLevel.All)),
  layer,
);



describe('Route - Message', (it) => {
  it.effect('when encoding', E.fn(function* () {
    const actual = Route.encodeMessageRoute({
      origin: 'Message',
      root  : 'root',
      hash  : 'hash',
      doken : {
        _tag     : 'Doken',
        id       : 'id',
        app_id   : 'app_id',
        token    : RDT.make('token'),
        ttl      : DateTime.unsafeMake(0),
        ephemeral: 0,
        type     : 0,
        status   : 'Active',
      },
    });

    expect(actual).toMatchSnapshot();
  }));

  it.effect('when decoding', E.fn(function* () {
    const actual = Route.decodeMessageRoute('dsx/root/0/0/id/0/2/token/hash');

    expect(actual).toMatchSnapshot();
  }));
});



describe('Route - Dialog', (it) => {
  it.effect('when encoding', E.fn(function* () {
    const actual = Route.encodeDialogRoute({
      origin: 'Message',
      root  : 'root',
      hash  : 'hash',
      doken : {
        _tag     : 'Doken',
        id       : 'id',
        app_id   : 'app_id',
        token    : RDT.make('token'),
        ttl      : DateTime.unsafeMake(0),
        ephemeral: 0,
        type     : 0,
        status   : 'Active',
      },
    });

    expect(actual).toMatchSnapshot();
  }));

  it.effect('when decoding', E.fn(function* () {
    const actual = Route.decodeDialogRoute('dsx/root/0/0/id/0');

    expect(actual).toMatchSnapshot();
  }));
});
