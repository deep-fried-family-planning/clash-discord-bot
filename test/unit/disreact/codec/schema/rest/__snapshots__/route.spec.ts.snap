// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Route - Dialog > when decoding 1`] = `
{
  "doken": {
    "_tag": "Doken",
    "ephemeral": 0,
    "id": "id",
    "status": "Active",
    "ttl": "1970-01-01T00:00:00.000Z",
    "type": 0,
  },
  "hash": "-",
  "origin": "Dialog",
  "root": "root",
}
`;

exports[`Route - Dialog > when encoding 1`] = `"dsx/root/0/0/id/0"`;

exports[`Route - Message > when decoding 1`] = `
{
  "doken": {
    "_tag": "Doken",
    "ephemeral": 0,
    "id": "id",
    "status": "Active",
    "token": "<redacted>",
    "ttl": "1970-01-01T00:00:00.000Z",
    "type": 0,
  },
  "hash": "hash",
  "origin": "Message",
  "root": "root",
}
`;

exports[`Route - Message > when encoding 1`] = `"dsx/root/0/0/id/0/2/token/hash"`;
